In [156]:
import pandas as pd
import numpy as np

In [157]:
credits = pd.read_csv('credits.csv')
movies = pd.read_csv('movies_metadata.csv')
users = pd.read_csv('users.csv')
ratings = pd.read_csv('ratings.csv')

C:\Users\HARSH\AppData\Local\Temp\ipykernel_4752\616419800.py:2: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  movies = pd.read_csv('movies_metadata.csv')


In [158]:
users.head(1)

userid   fname      lname   username gender            email password  \
0       1  Rashad  Castaneda  Rashad@61   Male  Salvador@et.com  6fVt6H@   

  confirmpassword  
0         6fVt6H@

In [159]:
movies.shape

(45466, 24)

In [160]:
movies.isnull().sum()

adult                        0
belongs_to_collection    40972
budget                       0
genres                       0
homepage                 37684
id                           0
imdb_id                     17
original_language           11
original_title               0
overview                   954
popularity                   5
poster_path                386
production_companies         3
production_countries         3
release_date                87
revenue                      6
runtime                    263
spoken_languages             6
status                      87
tagline                  25054
title                        6
video                        6
vote_average                 6
vote_count                   6
dtype: int64

In [161]:
credits['id']=credits['id'].astype(str)

In [162]:
movies=movies.merge(credits,on='id')

In [163]:
movies.head(1)

adult                              belongs_to_collection    budget  \
0  FALSE  [{'id': 10194, 'name': 'Toy Story Collection',...  30000000   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   

                               homepage   id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story  862  tt0114709                en   

  original_title                                           overview  ...  \
0      Toy Story  Led by Woody, Andy's toys live happily in his ...  ...   

  runtime                          spoken_languages    status tagline  \
0    81.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released     NaN   

       title  video  vote_average vote_count  \
0  Toy Story  False           7.7     5415.0   

                                                cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   

                                                crew  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...  

[1 rows x 26 columns]

In [164]:
movies = movies[['belongs_to_collection','genres','id','original_language','release_date','original_title','runtime','overview','cast','crew']]

In [165]:
movies.head(1)

belongs_to_collection  \
0  [{'id': 10194, 'name': 'Toy Story Collection',...   

                                              genres   id original_language  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...  862                en   

  release_date original_title  runtime  \
0   30-10-1995      Toy Story     81.0   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   

                                                cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   

                                                crew  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...

In [166]:
movies.isnull().sum()

belongs_to_collection    41038
genres                       0
id                           0
original_language           11
release_date                87
original_title               0
runtime                    260
overview                   954
cast                         0
crew                         0
dtype: int64

In [167]:
movies.duplicated().sum()

99

In [168]:
movies.dropna(inplace=True)

In [169]:
movies.duplicated().sum()

11

In [170]:
movies.drop_duplicates()

belongs_to_collection  \
0      [{'id': 10194, 'name': 'Toy Story Collection',...   
2      [{'id': 119050, 'name': 'Grumpy Old Men Collec...   
4      [{'id': 96871, 'name': 'Father of the Bride Co...   
9      [{'id': 645, 'name': 'James Bond Collection', ...   
12     [{'id': 117693, 'name': 'Balto Collection', 'p...   
...                                                  ...   
45427  [{'id': 37261, 'name': 'The Carry On Collectio...   
45430  [{'id': 37261, 'name': 'The Carry On Collectio...   
45441  [{'id': 37261, 'name': 'The Carry On Collectio...   
45443  [{'id': 477208, 'name': 'DC Super Hero Girls C...   
45454  [{'id': 200641, 'name': 'Red Lotus Collection'...   

                                                  genres      id  \
0      [{'id': 16, 'name': 'Animation'}, {'id': 35, '...     862   
2      [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   15602   
4                         [{'id': 35, 'name': 'Comedy'}]   11862   
9      [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...     710   
12     [{'id': 10751, 'name': 'Family'}, {'id': 16, '...   21032   
...                                                  ...     ...   
45427                     [{'id': 35, 'name': 'Comedy'}]   24568   
45430                     [{'id': 35, 'name': 'Comedy'}]   19307   
45441                     [{'id': 35, 'name': 'Comedy'}]   21251   
45443                  [{'id': 16, 'name': 'Animation'}]  460135   
45454  [{'id': 28, 'name': 'Action'}, {'id': 10769, '...  182030   

      original_language release_date                         original_title  \
0                    en   30-10-1995                              Toy Story   
2                    en   22-12-1995                       Grumpier Old Men   
4                    en   10-02-1995            Father of the Bride Part II   
9                    en   16-11-1995                              GoldenEye   
12                   en   22-12-1995                                  Balto   
...                 ...          ...                                    ...   
45427                en   01-09-1967                      Follow That Camel   
45430                en   29-05-1969                       Carry On Camping   
45441                en   01-10-1976                       Carry On England   
45443                en   30-08-2017  LEGO DC Super Hero Girls: Brain Drain   
45454                zh   22-12-1965                             火燒紅蓮寺之鴛鴦劍俠   

       runtime                                           overview  \
0         81.0  Led by Woody, Andy's toys live happily in his ...   
2        101.0  A family wedding reignites the ancient feud be...   
4        106.0  Just when George Banks has recovered from his ...   
9        130.0  James Bond must unmask the mysterious head of ...   
12        78.0  An outcast half-wolf risks his life to prevent...   
...        ...                                                ...   
45427     95.0  Bertram Oliphant West (also known as Bo West) ...   
45430     88.0  Sid and Bernie keep having their amorous inten...   
45441     89.0  Captain S. Melly takes over as the new Command...   
45443      0.0  When Supergirl, Wonder Woman, Batgirl, Bumbleb...   
45454     88.0  Kwei Wu happens upon a kidnapping with his wif...   

                                                    cast  \
0      [{'cast_id': 14, 'character': 'Woody (voice)',...   
2      [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
4      [{'cast_id': 1, 'character': 'George Banks', '...   
9      [{'cast_id': 1, 'character': 'James Bond', 'cr...   
12     [{'cast_id': 1, 'character': 'Balto (voice)', ...   
...                                                  ...   
45427  [{'cast_id': 3, 'character': 'Sergeant Nocker'...   
45430  [{'cast_id': 3, 'character': 'Sid Boggle', 'cr...   
45441  [{'cast_id': 4, 'character': 'Captain S. Melly...   
45443  [{'cast_id': 0, 'character': 'Wonder Woman', '...   
45454  [{'cast_id': 4, 'character': 'Gui Wu', 'credit...   

In [171]:
movies.shape

(4438, 10)

In [172]:
import ast

In [173]:
def convert(obj):
    L = []
    for i in ast.literal_eval(obj):
        L.append(i['name'])
    return L

In [174]:
movies['genres']=movies['genres'].apply(convert)

In [175]:
movies['cast']=movies['cast'].apply(convert)

In [176]:
def fetch_director(obj):
    L = []
    for i in ast.literal_eval(obj):
        if i['job'] == 'Director':
            L.append(i['name'])
            break
    return L

In [177]:
movies['crew']=movies['crew'].apply(fetch_director)

In [178]:
movies.head(1)

belongs_to_collection  \
0  [{'id': 10194, 'name': 'Toy Story Collection',...   

                        genres   id original_language release_date  \
0  [Animation, Comedy, Family]  862                en   30-10-1995   

  original_title  runtime                                           overview  \
0      Toy Story     81.0  Led by Woody, Andy's toys live happily in his ...   

                                                cast             crew  
0  [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...  [John Lasseter]

In [179]:
movies.iloc[0].belongs_to_collection	

"[{'id': 10194, 'name': 'Toy Story Collection', 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg', 'backdrop_path': '/9FBwqcd9IRruEDUrTdcaafOMKUq.jpg'}]"

In [180]:
movies.head(1)

belongs_to_collection  \
0  [{'id': 10194, 'name': 'Toy Story Collection',...   

                        genres   id original_language release_date  \
0  [Animation, Comedy, Family]  862                en   30-10-1995   

  original_title  runtime                                           overview  \
0      Toy Story     81.0  Led by Woody, Andy's toys live happily in his ...   

                                                cast             crew  
0  [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...  [John Lasseter]

In [181]:
import ast
def convert1(obj):
    L = []
    for i in ast.literal_eval(obj):
        L.append(i['poster_path'])
        L.append(i['backdrop_path'])
    return L

In [182]:
movies['belongs_to_collection']=movies['belongs_to_collection'].apply(convert1)

In [183]:
movies.head(1)

belongs_to_collection  \
0  [/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg, /9FBwqcd9IR...   

                        genres   id original_language release_date  \
0  [Animation, Comedy, Family]  862                en   30-10-1995   

  original_title  runtime                                           overview  \
0      Toy Story     81.0  Led by Woody, Andy's toys live happily in his ...   

                                                cast             crew  
0  [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...  [John Lasseter]

In [184]:
movies2 = movies[['genres','id','original_language','release_date','original_title','runtime','overview','cast','crew']]

In [185]:
movies2.head(1)

genres   id original_language release_date  \
0  [Animation, Comedy, Family]  862                en   30-10-1995   

  original_title  runtime                                           overview  \
0      Toy Story     81.0  Led by Woody, Andy's toys live happily in his ...   

                                                cast             crew  
0  [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...  [John Lasseter]

In [186]:
movies3 = movies[['belongs_to_collection','id']]

In [187]:
movies3.head(1)

belongs_to_collection   id
0  [/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg, /9FBwqcd9IR...  862

In [188]:
movies3[['image1','image2']] = pd.DataFrame(movies3.belongs_to_collection.tolist(), index= movies3.index)

C:\Users\HARSH\AppData\Local\Temp\ipykernel_4752\318437437.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies3[['image1','image2']] = pd.DataFrame(movies3.belongs_to_collection.tolist(), index= movies3.index)
C:\Users\HARSH\AppData\Local\Temp\ipykernel_4752\318437437.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies3[['image1','image2']] = pd.DataFrame(movies3.belongs_to_collection.tolist(), index= movies3.index)


In [189]:
movies3.head(1)

belongs_to_collection   id  \
0  [/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg, /9FBwqcd9IR...  862   

                             image1                            image2  
0  /7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg  /9FBwqcd9IRruEDUrTdcaafOMKUq.jpg

In [190]:
movies3 = movies3[['id','image1','image2']]

In [191]:
movies3.head(1)

id                            image1                            image2
0  862  /7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg  /9FBwqcd9IRruEDUrTdcaafOMKUq.jpg

In [192]:
movies=movies.merge(movies3,on='id')

In [193]:
movies.head(1)

belongs_to_collection  \
0  [/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg, /9FBwqcd9IR...   

                        genres   id original_language release_date  \
0  [Animation, Comedy, Family]  862                en   30-10-1995   

  original_title  runtime                                           overview  \
0      Toy Story     81.0  Led by Woody, Andy's toys live happily in his ...   

                                                cast             crew  \
0  [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...  [John Lasseter]   

                             image1                            image2  
0  /7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg  /9FBwqcd9IRruEDUrTdcaafOMKUq.jpg

In [194]:
final_movies = movies[['genres','id','original_language','release_date','original_title','runtime','overview','cast','crew','image1','image2']]

In [195]:
final_movies.head(1)

genres   id original_language release_date  \
0  [Animation, Comedy, Family]  862                en   30-10-1995   

  original_title  runtime                                           overview  \
0      Toy Story     81.0  Led by Woody, Andy's toys live happily in his ...   

                                                cast             crew  \
0  [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...  [John Lasseter]   

                             image1                            image2  
0  /7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg  /9FBwqcd9IRruEDUrTdcaafOMKUq.jpg

In [196]:
final_movies['runtime'] = final_movies['runtime'].astype(str)          
final_movies['runtime'] = final_movies['runtime'].str.replace('.0', ' ')

C:\Users\HARSH\AppData\Local\Temp\ipykernel_4752\318855864.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_movies['runtime'] = final_movies['runtime'].astype(str)
C:\Users\HARSH\AppData\Local\Temp\ipykernel_4752\318855864.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_movies['runtime'] = final_movies['runtime'].str.replace('.0', ' ')


In [197]:
final_movies.head(1)

genres   id original_language release_date  \
0  [Animation, Comedy, Family]  862                en   30-10-1995   

  original_title runtime                                           overview  \
0      Toy Story     81   Led by Woody, Andy's toys live happily in his ...   

                                                cast             crew  \
0  [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...  [John Lasseter]   

                             image1                            image2  
0  /7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg  /9FBwqcd9IRruEDUrTdcaafOMKUq.jpg

In [198]:
content_list = final_movies

In [199]:
content_list.head(1)

genres   id original_language release_date  \
0  [Animation, Comedy, Family]  862                en   30-10-1995   

  original_title runtime                                           overview  \
0      Toy Story     81   Led by Woody, Andy's toys live happily in his ...   

                                                cast             crew  \
0  [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...  [John Lasseter]   

                             image1                            image2  
0  /7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg  /9FBwqcd9IRruEDUrTdcaafOMKUq.jpg

In [200]:
final_movies['genres']=movies['genres'].str[0]
final_movies['crew']=movies['crew'].str[0]

C:\Users\HARSH\AppData\Local\Temp\ipykernel_4752\1366408633.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_movies['genres']=movies['genres'].str[0]
C:\Users\HARSH\AppData\Local\Temp\ipykernel_4752\1366408633.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_movies['crew']=movies['crew'].str[0]


In [201]:
final_movies.head(2)

genres     id original_language release_date    original_title runtime  \
0  Animation    862                en   30-10-1995         Toy Story     81    
1    Romance  15602                en   22-12-1995  Grumpier Old Men    101    

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  A family wedding reignites the ancient feud be...   

                                                cast           crew  \
0  [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...  John Lasseter   
1  [Walter Matthau, Jack Lemmon, Ann-Margret, Sop...  Howard Deutch   

                             image1                            image2  
0  /7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg  /9FBwqcd9IRruEDUrTdcaafOMKUq.jpg  
1  /nLvUdqgPgm3F85NMCii9gVFUcet.jpg  /hypTnLot2z8wpFS7qwsQHW1uV8u.jpg

In [202]:
final_movies['release_date'] = pd.to_datetime(final_movies['release_date'])

C:\Users\HARSH\AppData\Local\Temp\ipykernel_4752\2262031045.py:1: UserWarning: Parsing dates in %d-%m-%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  final_movies['release_date'] = pd.to_datetime(final_movies['release_date'])
C:\Users\HARSH\AppData\Local\Temp\ipykernel_4752\2262031045.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_movies['release_date'] = pd.to_datetime(final_movies['release_date'])


In [203]:
final_movies['release_date'] = final_movies['release_date'].dt.year

C:\Users\HARSH\AppData\Local\Temp\ipykernel_4752\3825791924.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_movies['release_date'] = final_movies['release_date'].dt.year


In [204]:
final_movies.head(1)

genres   id original_language  release_date original_title runtime  \
0  Animation  862                en          1995      Toy Story     81    

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   

                                                cast           crew  \
0  [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...  John Lasseter   

                             image1                            image2  
0  /7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg  /9FBwqcd9IRruEDUrTdcaafOMKUq.jpg

In [205]:
ratings = ratings[['userId','movieId','rating']]

In [206]:
ratings['id']=ratings['movieId']

In [207]:
ratings = ratings[['userId','id','rating']]

In [208]:
ratings.head(1)

userId   id  rating
0       1  110     1.0

In [209]:
ratings.shape

(1048575, 3)

In [210]:
ratings['id']=ratings['id'].astype(str)

In [211]:
ratings_with_id=ratings.merge(final_movies,on='id')

In [212]:
ratings_with_id.head(1)

userId   id  rating genres original_language  release_date  \
0       1  110     1.0  Drama                fr          1994   

           original_title runtime  \
0  Trois couleurs : Rouge     99    

                                            overview  \
0  Red This is the third film from the trilogy by...   

                                                cast                  crew  \
0  [Irène Jacob, Jean-Louis Trintignant, Jean-Pie...  Krzysztof Kieślowski   

                             image1                            image2  
0  /rVdd23QuT5rHX7lZvuAkRRUkeCZ.jpg  /AeHExfHIl70SZCea907KfEoSkfJ.jpg

In [213]:
ratings_with_id.isnull().sum()

userId                  0
id                      0
rating                  0
genres                 18
original_language       0
release_date            0
original_title          0
runtime                 0
overview                0
cast                    0
crew                    4
image1               6554
image2               9617
dtype: int64

In [214]:
num_rating_df=ratings_with_id.groupby('id').count()['rating'].reset_index()
num_rating_df.rename(columns={'rating':'num_ratings'},inplace=True)
num_rating_df

id  num_ratings
0    100042            1
1    100450            5
2    102899            1
3    103751            1
4       105          423
..      ...          ...
762   98344            1
763     984            9
764     990           41
765     993            9
766   99861            2

[767 rows x 2 columns]

In [215]:
avg_rating_df=ratings_with_id.groupby('id')['rating'].mean().rename('avg_rating')
avg_rating_df

id
100042    3.000000
100450    3.100000
102899    3.000000
103751    4.500000
105       3.345154
            ...   
98344     3.000000
984       2.666667
990       2.585366
993       2.444444
99861     3.250000
Name: avg_rating, Length: 767, dtype: float64

In [216]:
popular_df = num_rating_df.merge(avg_rating_df,on='id')
popular_df

id  num_ratings  avg_rating
0    100042            1    3.000000
1    100450            5    3.100000
2    102899            1    3.000000
3    103751            1    4.500000
4       105          423    3.345154
..      ...          ...         ...
762   98344            1    3.000000
763     984            9    2.666667
764     990           41    2.585366
765     993            9    2.444444
766   99861            2    3.250000

[767 rows x 3 columns]

# TOP 20 HIGHEST RATED 

In [217]:
popular_df = popular_df[popular_df['num_ratings']>=100].sort_values('avg_rating',ascending=False).head(21)
popular_df

id  num_ratings  avg_rating
161  2019          539    4.286642
488   527         2733    4.261983
31   1284          238    4.224790
300   296         3464    4.199480
21   1248          206    4.174757
752   950          154    4.162338
753   951          177    4.118644
309   306          314    4.116242
740   923          827    4.100967
564   608         2073    4.094067
22   1250          568    4.088908
29   1281          179    4.069832
23   1251          155    4.051613
756   955          179    4.047486
742   928          214    4.035047
25   1266          629    4.034976
11    110         2594    4.025251
754   953          677    4.015510
373  3683          221    3.997738
755   954          271    3.996310
419  4327          192    3.994792

In [218]:
popular_df = popular_df.merge(final_movies,on='id').drop_duplicates('id')[['id','original_title','genres','overview','cast','crew','avg_rating','image1','image2','runtime','release_date','original_language']]

In [219]:
print(popular_df)

      id                      original_title     genres  \
0   2019                         Hard Target     Action   
1    527                  Once Were Warriors      Drama   
2   1284      Torrente 2: Misión en Marbella     Comedy   
3    296  Terminator 3: Rise of the Machines     Action   
4   1248                     Hannibal Rising      Crime   
5    950               Ice Age: The Meltdown  Animation   
6    951                    Kindergarten Cop     Comedy   
7    306               Beverly Hills Cop III     Action   
8    923                    Dawn of the Dead     Horror   
9    608                     Men in Black II     Action   
10  1250                         Ghost Rider   Thriller   
11  1281                                Bean     Comedy   
12  1251               Letters from Iwo Jima     Action   
13   955              Mission: Impossible II  Adventure   
14   928           Gremlins 2: The New Batch     Comedy   
15  1266                        Street Kings     Action 

In [220]:
popular_df.head(1)

id original_title  genres  \
0  2019    Hard Target  Action   

                                            overview  \
0  When a woman's father goes missing, she enlist...   

                                                cast      crew  avg_rating  \
0  [Jean-Claude Van Damme, Arnold Vosloo, Lance H...  John Woo    4.286642   

                             image1                            image2 runtime  \
0  /g7Bc4ik16xOBz21BQOcsxrBL8BO.jpg  /wcBIwQY888HX1QWpnv43lAujoId.jpg     97    

   release_date original_language  
0          1993                en

In [221]:
popular_df.isnull().sum()

id                   0
original_title       0
genres               0
overview             0
cast                 0
crew                 0
avg_rating           0
image1               1
image2               1
runtime              0
release_date         0
original_language    0
dtype: int64

In [222]:
popular_df.dropna(inplace=True)

In [223]:
popular_df.isnull().sum()

id                   0
original_title       0
genres               0
overview             0
cast                 0
crew                 0
avg_rating           0
image1               0
image2               0
runtime              0
release_date         0
original_language    0
dtype: int64

In [224]:
print(popular_df)

      id                      original_title     genres  \
0   2019                         Hard Target     Action   
2   1284      Torrente 2: Misión en Marbella     Comedy   
3    296  Terminator 3: Rise of the Machines     Action   
4   1248                     Hannibal Rising      Crime   
5    950               Ice Age: The Meltdown  Animation   
6    951                    Kindergarten Cop     Comedy   
7    306               Beverly Hills Cop III     Action   
8    923                    Dawn of the Dead     Horror   
9    608                     Men in Black II     Action   
10  1250                         Ghost Rider   Thriller   
11  1281                                Bean     Comedy   
12  1251               Letters from Iwo Jima     Action   
13   955              Mission: Impossible II  Adventure   
14   928           Gremlins 2: The New Batch     Comedy   
15  1266                        Street Kings     Action   
16   110              Trois couleurs : Rouge      Drama 

In [225]:
import pickle
pickle.dump(popular_df,open('populardf.pkl','wb'))

# BASED ON GENRE

In [226]:
final_movies.head(1)

genres   id original_language  release_date original_title runtime  \
0  Animation  862                en          1995      Toy Story     81    

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   

                                                cast           crew  \
0  [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...  John Lasseter   

                             image1                            image2  
0  /7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg  /9FBwqcd9IRruEDUrTdcaafOMKUq.jpg

In [227]:
ratings_with_id.head(1)

userId   id  rating genres original_language  release_date  \
0       1  110     1.0  Drama                fr          1994   

           original_title runtime  \
0  Trois couleurs : Rouge     99    

                                            overview  \
0  Red This is the third film from the trilogy by...   

                                                cast                  crew  \
0  [Irène Jacob, Jean-Louis Trintignant, Jean-Pie...  Krzysztof Kieślowski   

                             image1                            image2  
0  /rVdd23QuT5rHX7lZvuAkRRUkeCZ.jpg  /AeHExfHIl70SZCea907KfEoSkfJ.jpg

In [228]:
avg_rating_df.head(1)

id
100042    3.0
Name: avg_rating, dtype: float64

In [229]:
movie_list = num_rating_df.merge(avg_rating_df,on='id')
movie_list

id  num_ratings  avg_rating
0    100042            1    3.000000
1    100450            5    3.100000
2    102899            1    3.000000
3    103751            1    4.500000
4       105          423    3.345154
..      ...          ...         ...
762   98344            1    3.000000
763     984            9    2.666667
764     990           41    2.585366
765     993            9    2.444444
766   99861            2    3.250000

[767 rows x 3 columns]

In [230]:
movie_list = movie_list.merge(final_movies,on='id').drop_duplicates('id')[['id','original_title','genres','overview','cast','crew','avg_rating','image1','image2','runtime','release_date','original_language']]

In [231]:
movie_list.shape

(767, 12)

# top drama

In [232]:
top_drama = movie_list[movie_list['genres']=='Drama'].sort_values('avg_rating',ascending=False).head(20)
top_drama

id                     original_title genres  \
68    1555                L'Auberge espagnole  Drama   
503    527                 Once Were Warriors  Drama   
663  78140      Billy Jack Goes to Washington  Drama   
319   3035                       Frankenstein  Drama   
347  31374                       続宮本武蔵　一乗寺の決闘  Drama   
11     110             Trois couleurs : Rouge  Drama   
607    670                               올드보이  Drama   
561   5817                         Du levande  Drama   
480   4978                   An American Tail  Drama   
20    1246                       Rocky Balboa  Drama   
534   5608  Die Nibelungen, Teil 1: Siegfried  Drama   
325    307                  Roma città aperta  Drama   
98      16                 Dancer in the Dark  Drama   
666     80                      Before Sunset  Drama   
367  33660                             Shiloh  Drama   
500   5203                        Wild Orchid  Drama   
244  25934                  To Sir, with Love  Drama   
582  61108                       Emmanuelle 6  Drama   
158   1931                     Stomp the Yard  Drama   
160   1951                          Manderlay  Drama   

                                              overview  \
68   A strait-laced French student moves into an ap...   
503  A drama about a Maori family lving in Auckland...   
663  After a senator suddenly dies after completing...   
319  Henry Frankenstein is a doctor who is trying t...   
347  After years on the road establishing his reput...   
11   Red This is the third film from the trilogy by...   
607  With no clue how he came to be imprisoned, dru...   
561  In the Swedish city of Lethe, people from diff...   
480  A young mouse named Fievel and his family deci...   
20   When he loses a highly publicized virtual boxi...   
534  A young hero defeats a dragon to find acceptan...   
325  Open City is a landmark in film history. Filme...   
98   Selma, a Czech immigrant on the verge of blind...   
666  Nine years ago two strangers met by chance and...   
367  An abused beagle runs away from his owner. On ...   
500  A woman lawyer becomes mesmerized by a self-ma...   
244  Idealistic engineer-trainee and his experience...   
582  Journey with Emmanuelle and the world's most e...   
158  After the death of his younger brother, a trou...   
160  In 1933, after leaving Dogville, Grace Margare...   

                                                  cast                  crew  \
68   [Romain Duris, Audrey Tautou, Kelly Reilly, Cé...       Cédric Klapisch   
503  [Rena Owen, Temuera Morrison, Mamaengaroa Kerr...          Lee Tamahori   
663  [Tom Laughlin, Delores Taylor, E.G. Marshall, ...          Tom Laughlin   
319  [Colin Clive, Mae Clarke, John Boles, Boris Ka...           James Whale   
347  [Toshirō Mifune, Koji Tsuruta, Kaoru Yachigusa...       Hiroshi Inagaki   
11   [Irène Jacob, Jean-Louis Trintignant, Jean-Pie...  Krzysztof Kieślowski   
607  [Choi Min-sik, Yoo Ji-tae, Kang Hye-jung, Kim ...        Park Chan-wook   
561  [Håkan Angser, Eric Bäckman, Patrik Anders Edg...         Roy Andersson   
480  [Phillip Glasser, Dom DeLuise, Nehemiah Persof...             Don Bluth   
20   [Sylvester Stallone, Burt Young, Milo Ventimig...    Sylvester Stallone   
534  [Uwe Beyer, Karin Dor, Maria Marlow, Rolf Henn...          Harald Reinl   
325  [Aldo Fabrizi, Anna Magnani, Marcello Pagliero...    Roberto Rossellini   
98   [Björk, Catherine Deneuve, David Morse, Peter ...        Lars von Trier   
666  [Ethan Hawke, Julie Delpy, Vernon Dobtcheff, L...     Richard Linklater   
367  [Blake Heron, J. Madison Wright, Michael Moria...       Dale Rosenbloom   
500  [Jacqueline Bisset, Carré Otis, Mickey Rourke,...           Zalman King   
244  [Sidney Poitier, Suzy Kendall, Christian Rober...         James Clavell   
582  [Natalie Uher, Jean-René Gossart, Tamira, Gust...         Bruno Zincone   
158  [Columbus Short, Meagan Good, Ne-Yo, Darrin He...         Sylvain White   
160  [Bryce 

In [233]:
top_drama.isnull().sum()
top_drama.dropna(inplace=True)
top_drama.isnull().sum()

id                   0
original_title       0
genres               0
overview             0
cast                 0
crew                 0
avg_rating           0
image1               0
image2               0
runtime              0
release_date         0
original_language    0
dtype: int64

In [234]:
import pickle
pickle.dump(top_drama,open('top_drama.pkl','wb'))

# top action movie

In [235]:
top_action = movie_list[movie_list['genres']=='Action'].sort_values('avg_rating',ascending=False).head(11)
top_action

id                               original_title  genres  \
520  54445                          S.W.A.T.: Firefight  Action   
176   2019                                  Hard Target  Action   
207   2330                                         Taxi  Action   
620   6935  Yamakasi - Les samouraïs des temps modernes  Action   
485   5056                       Deadlier Than the Male  Action   
372  34326                                       仁義なき戦い  Action   
497   5174                                  Rush Hour 3  Action   
315    296           Terminator 3: Rise of the Machines  Action   
324    306                        Beverly Hills Cop III  Action   
579    608                              Men in Black II  Action   
23    1251                        Letters from Iwo Jima  Action   

                                              overview  \
520  Los Angeles S.W.A.T. officer, Lt. Paul Cutler,...   
176  When a woman's father goes missing, she enlist...   
207  In Marseilles (France), skilled pizza delivery...   
620  Yamakasi - Les samouraïs des temps modernes is...   
485  In this 1967 drama, resourceful British agent ...   
372  In the teeming black markets of postwar Japan,...   
497  After an attempted assassination on Ambassador...   
315  It's been 10 years since John Connor saved Ear...   
324  Back in sunny southern California and on the t...   
579  Kay and Jay reunite to provide our best, last ...   
23   The story of the battle of Iwo Jima between th...   

                                                  cast              crew  \
520  [Gabriel Macht, Robert Patrick, Carly Pope, Gi...        Benny Boom   
176  [Jean-Claude Van Damme, Arnold Vosloo, Lance H...          John Woo   
207  [Samy Naceri, Frédéric Diefenthal, Marion Coti...      Gérard Pirès   
620  [Châu Belle Dinh, Williams Belle, Malik Diouf,...     Ariel Zeitoun   
485  [Richard Johnson, Elke Sommer, Sylva Koscina, ...      Ralph Thomas   
372  [Bunta Sugawara, Hiroki Matsukata, Kunie Tanak...    Kinji Fukasaku   
497  [Chris Tucker, Jackie Chan, Hiroyuki Sanada, M...      Brett Ratner   
315  [Arnold Schwarzenegger, Nick Stahl, Claire Dan...   Jonathan Mostow   
324  [Eddie Murphy, Judge Reinhold, Héctor Elizondo...       John Landis   
579  [Tommy Lee Jones, Will Smith, Rip Torn, Lara F...  Barry Sonnenfeld   
23   [Ken Watanabe, Kazunari Ninomiya, Tsuyoshi Iha...    Clint Eastwood   

     avg_rating                            image1  \
520    5.000000  /iaBn9GPUik9xwfHIelzTXJi5ClY.jpg   
176    4.286642  /g7Bc4ik16xOBz21BQOcsxrBL8BO.jpg   
207    4.225000  /nCzy28TbMJDJLxuaPgQZrLgmuT7.jpg   
620    4.208333  /6t7QBvkMq02vgGuwxH7VduDbZ4j.jpg   
485    4.205882                              None   
372    4.200000  /3r2EkL4OpBB5iwN52C5yNdbF1NU.jpg   
497    4.200000  /pTjvzlYByJuZ4TTkmlHIe4aDFy2.jpg   
315    4.199480  /vxiKtcxAJxHhlg2H1X8y7zcM3k6.jpg   
324    4.116242  /fjKbHh7GpTARbiLj0RF7zc5iD0G.jpg   
579    4.094067  /rWcgZaoJ0nxTFRn0H6w9bdtoDUI.jpg   
23     4.051613  /rnRE6JGz9yjYUAbWIHYkGu9B8pj.jpg   

                               image2 runtime  release_date original_language  
520  /75To7hSFaAuciz4EXAFfj4LH6Od.jpg     89           2011                en  
176  /wcBIwQY888HX1QWpnv43lAujoId.jpg     97           1993                en  
207  /74R8bq4RDuxK3F0VPGlse4YH3Ab.jpg     86           1998                fr  
620  /brTwRwoszOadojCoRzCbNjbYd9S.jpg     90           2001                en  
485                              None    101           1967                en  
372                              None     99           1973                ja  
497  /uoDt3eCCKpcxSQVWjPTgN8T0gjm.jpg     91           2007                en  
315  /cpmbkwSxZnKO69V82A9a34tvk2E.jpg    109           2003                en  
324  /lhZQuT76B9YcxZIudXZMhhWCKcl.jpg    104           1994                en  
579  /xKz6YaOvmnIJQ3MYIpnaa7FBaAB.jpg     88           2002                en  
23   /dIYpFavJ8vNCv9pefEabsUL1pDP.jpg    141           2006                e

In [236]:
top_action.isnull().sum()
top_action.dropna(inplace=True)
top_action.isnull().sum()

id                   0
original_title       0
genres               0
overview             0
cast                 0
crew                 0
avg_rating           0
image1               0
image2               0
runtime              0
release_date         0
original_language    0
dtype: int64

In [237]:
import pickle
pickle.dump(top_action,open('top_action.pkl','wb'))

# top crime

In [238]:
top_crime = movie_list[movie_list['genres']=='Crime'].sort_values('avg_rating',ascending=False).head(20)
top_crime

id                          original_title genres  \
378    3485                 The Castle of Fu Manchu  Crime   
442   44674                            Après la vie  Crime   
21     1248                         Hannibal Rising  Crime   
469   47404           Nick Carter, Master Detective  Crime   
247   26005                                    香港製造  Crime   
526     553                                Dogville  Crime   
332    3093                        Basic Instinct 2  Crime   
317     298                        Ocean's Thirteen  Crime   
542   56508    Fantômas: À l'ombre de la guillotine  Crime   
614   68163                           The Hidden II  Crime   
544   56511          Fantômas: Juve contre Fantômas  Crime   
739    8923                  Green Street Hooligans  Crime   
709    8617                              Prom Night  Crime   
217     242                 The Godfather: Part III  Crime   
178    2043                     Along Came a Spider  Crime   
105    1642                                 The Net  Crime   
200    2262         Die Sehnsucht der Veronika Voss  Crime   
373     344   Bang Boom Bang - Ein todsicheres Ding  Crime   
118  168056  Fjällbackamorden 05 - Vänner för livet  Crime   
386   36576               Curse of the Pink Panther  Crime   

                                              overview  \
378  The evil mastermind Fu Manchu plots his latest...   
442  The final installment in director Lucas Belvau...   
21   The story of the early, murderous roots of the...   
469  Detective Nick Carter is brought in to foil sp...   
247  The first independent film to emerge from the ...   
526  A barren soundstage is stylishly utilized to c...   
332  Novelist Catherine Tramell is once again in tr...   
317  Danny Ocean's team of criminals are back and c...   
542  For the arch-criminal Fantômas, no crime is to...   
614  The alien criminal from the first movie is dea...   
544  When a disfigured body of a woman is found in ...   
739  After being wrongfully expelled from Harvard U...   
709  Donna's senior prom is supposed to be the best...   
217  In the midst of trying to legitimize his busin...   
178  When a teacher kidnaps a girl from a prestigio...   
105  Angela Bennett is a freelance software enginee...   
200  In Munich in 1955, sports reporter Robert Kroh...   
373  Bank robber Kelle Grabowski escapes from priso...   
118  The successful detective writer, Erica Falck, ...   
386  Inspector Clouseau disappears, and the Surete ...   

                                                  cast  \
378  [Christopher Lee, Richard Greene, Howard Mario...   
442  [Dominique Blanc, Gilbert Melki, Ornella Muti,...   
21   [Gaspard Ulliel, Aaran Thomas, Gong Li, Domini...   
469  [Walter Pidgeon, Rita Johnson, Henry Hull, Sta...   
247  [Sam Lee, Neiky Yim Hui Chi, Wenders Li, Carol...   
526  [Nicole Kidman, Harriet Andersson, Lauren Baca...   
332  [Sharon Stone, David Morrissey, Charlotte Ramp...   
317  [George Clooney, Brad Pitt, Matt Damon, Al Pac...   
542  [René Navarre, Georges Melchior, Edmund Breon,...   
614  [Raphael Sbarge, Juliet Beck, Stanton, Tony Th...   
544  [René Navarre, Georges Melchior, Edmund Breon,...   
739  [Elijah Wood, Claire Forlani, Charlie Hunnam, ...   
709  [Brittany Snow, Scott Porter, Jessica Stroup, ...   
217  [Al Pacino, Diane Keaton, Andy García, Talia S...   
178  [Morgan Freeman, Monica Potter, Michael Wincot...   
105  [Sandra Bullock, Jeremy Northam, Dennis Miller...   
200  [Rosel Zech, Hilmar Thate, Annemarie Düringer,...   
373  [Oliver Korittke, Markus Knüfken, Ralf Richter...   
118  [Claudia Galli, Richard Ulfsäter, Eva Fritjofs...   
386  [David Niven, Robert Wagner, Herbert Lom, Joan...   

                         crew  avg_rating                            image1  \
378              Jesús Franco    4.500000  /hu15Dmis3D9mpT0P7Ni7kdJpBaF.jpg   
442             Lucas Belvaux    4.500000  /hiY1IaFlHf4D1EDAYtHBtAYraIa.jpg   
21               Peter Webber  

In [239]:
top_crime.isnull().sum()
top_crime.dropna(inplace=True)
top_crime.isnull().sum()

id                   0
original_title       0
genres               0
overview             0
cast                 0
crew                 0
avg_rating           0
image1               0
image2               0
runtime              0
release_date         0
original_language    0
dtype: int64

In [240]:
import pickle
pickle.dump(top_crime,open('top_crime.pkl','wb'))

# top comdies

In [241]:
top_comedy = movie_list[movie_list['genres']=='Comedy'].sort_values('avg_rating',ascending=False).head(10)
top_comedy

id                  original_title  genres  \
239   25833     Eating Out: All You Can Eat  Comedy   
551    5721                          Vixen!  Comedy   
701    8491             Weekend at Bernie's  Comedy   
31     1284  Torrente 2: Misión en Marbella  Comedy   
768     951                Kindergarten Cop  Comedy   
29     1281                            Bean  Comedy   
757     928       Gremlins 2: The New Batch  Comedy   
639   71995                 Risate di gioia  Comedy   
80   159447            С новым годом, мамы!  Comedy   
619     693                Meet the Fockers  Comedy   

                                              overview  \
239  Tiffani and her friend Casey try to lure the g...   
551  Vixen lives in a Canadian mountain resort with...   
701  Two friends are invited for a weekend to a lux...   
31   For this second film in the cult comedy series...   
768  Hard-edged cop John Kimble gets more than he b...   
29   Bean works as a caretaker at Britain's formida...   
757  Young sweethearts Billy and Kate move to the B...   
639  On New Year's Eve, an insecure, struggling act...   
80   A continuation of a famous Russian comedy "Mam...   
619  Hard-to-crack ex-CIA man, Jack Byrnes and his ...   

                                                  cast              crew  \
239  [Rebekah Kochan, Daniel Skelton, Chris Salvato...     Glenn Gaylord   
551  [Erica Gavin, Garth Pillsbury, Harrison Page, ...        Russ Meyer   
701  [Andrew McCarthy, Jonathan Silverman, Catherin...      Ted Kotcheff   
31   [Santiago Segura, Gabino Diego, Tony Leblanc, ...   Santiago Segura   
768  [Arnold Schwarzenegger, Penelope Ann Miller, P...      Ivan Reitman   
29   [Rowan Atkinson, Peter MacNicol, John Mills, P...         Mel Smith   
757  [Zach Galligan, Phoebe Cates, John Glover, Rob...         Joe Dante   
639  [Anna Magnani, Totò, Ben Gazzara, Fred Clark, ...   Mario Monicelli   
80   [Alain Delon, Elizaveta Boyarskaya, Agniya Dit...  Sarik Andreasyan   
619  [Ben Stiller, Teri Polo, Robert De Niro, Dusti...         Jay Roach   

     avg_rating                            image1  \
239    4.416667  /uK86CFpVQjP7t4f7U9sIH8FotKN.jpg   
551    4.250000  /yWIVyrHMJkRl9zhxexZSUXyXtIN.jpg   
701    4.233333  /gJVBXVetIkgVVMbzODQ3dJwvQkV.jpg   
31     4.224790  /crCyDVyXc4kVyO2a0Tk5fposzm8.jpg   
768    4.118644  /q5Pk0utXHTqOgvRtlRxu8AMqGLC.jpg   
29     4.069832  /61Jcc1Gwc2KE7sIJ1HPqqdt67ji.jpg   
757    4.035047  /yiZ18XZHFNjFfEDnEkFCrHVHXcn.jpg   
639    4.000000  /4ayJsjC3djGwU9eCWUokdBWvdLC.jpg   
80     4.000000  /o567GZtOIOeXzRziRyF4gUZCWn7.jpg   
619    4.000000   /WFkinadVm4NRynRKuUcaLFNIDx.jpg   

                               image2 runtime  release_date original_language  
239                              None     80           2009                en  
551                              None     70           1968                en  
701  /khVvskKc4VzAHbHtQMhKGeKWDxC.jpg     97           1989                en  
31   /hm4yNDojfPUi34OKLEsHiweM0iT.jpg     99           2001                es  
768  /qdKZwhWbIjlKbBOWsHL1cCcGSa2.jpg    111           1990                en  
29   /aIXcqHz5VKRLXHPsNo0dFoK7YP5.jpg     85           1997                en  
757  /2Oz7EyiM9nNeP7appRFkMZNeYRo.jpg    106           1990                en  
639  /jaUuprubvAxXLAY5hUfrNjxccUh.jpg    106           1960                it  
80   /psFngVh40DpApbCEM4sE6svz65K.jpg     90           2012                ru  
619  /7UWRq79VHc58QXTSn57UsOUKQLZ.jpg    115           2004                en

In [242]:
import pickle
pickle.dump(top_comedy,open('top_comedy.pkl','wb'))

# top advanture

In [243]:
top_adventure = movie_list[movie_list['genres']=='Adventure'].sort_values('avg_rating',ascending=False).head(11)
top_adventure

id                                     original_title     genres  \
548   5683                            Pee-wee's Big Adventure  Adventure   
659    767             Harry Potter and the Half-Blood Prince  Adventure   
606    668                    On Her Majesty's Secret Service  Adventure   
771    955                             Mission: Impossible II  Adventure   
133   1734                                  The Mummy Returns  Adventure   
358  32657  Percy Jackson & the Olympians: The Lightning T...  Adventure   
770    954                                Mission: Impossible  Adventure   
557     58         Pirates of the Caribbean: Dead Man's Chest  Adventure   
152   1893          Star Wars: Episode I - The Phantom Menace  Adventure   
197   2202                                   Lassie Come Home  Adventure   
492   5086                         Quest for the Mighty Sword  Adventure   

                                              overview  \
548  The eccentric and childish Pee-wee Herman emba...   
659  As Harry begins his sixth year at Hogwarts, he...   
606  James Bond tracks archnemesis Ernst Blofeld to...   
771  With computer genius Luther Stickell at his si...   
133  Rick and Evelyn O'Connell, along with their 8 ...   
358  Accident prone teenager, Percy discovers he's ...   
770  When Ethan Hunt, the leader of a crack espiona...   
557  Captain Jack Sparrow works his way out of a bl...   
152  Anakin Skywalker, a young slave strong with th...   
197  Hard times came for Carraclough family and the...   
492  Once upon a time a god gave a mighty sword to ...   

                                                  cast             crew  \
548  [Paul Reubens, E.G. Daily, Mark Holton, Diane ...       Tim Burton   
659  [Daniel Radcliffe, Rupert Grint, Emma Watson, ...      David Yates   
606  [George Lazenby, Diana Rigg, Telly Savalas, Ga...    Peter R. Hunt   
771  [Tom Cruise, Dougray Scott, Thandie Newton, Vi...         John Woo   
133  [Brendan Fraser, Rachel Weisz, John Hannah, Ar...  Stephen Sommers   
358  [Logan Lerman, Brandon T. Jackson, Alexandra D...   Chris Columbus   
770  [Tom Cruise, Jon Voight, Emmanuelle Béart, Hen...   Brian De Palma   
557  [Johnny Depp, Orlando Bloom, Keira Knightley, ...   Gore Verbinski   
152  [Liam Neeson, Ewan McGregor, Natalie Portman, ...     George Lucas   
197  [Roddy McDowall, Donald Crisp, Dame May Whitty...   Fred M. Wilcox   
492  [Eric Allan Kramer, Margaret Lenzey, Donald O'...      Joe D'Amato   

     avg_rating                            image1  \
548    4.500000  /m806SvjQ9b6FPFEzS2ODG6lbXqk.jpg   
659    4.107143  /8Xmkc1HvCOpMlbFvVabrtr6HAsp.jpg   
606    4.094595   /HORpg5CSkmeQlAolx3bKMrKgfi.jpg   
771    4.047486  /gwyJPIhCK4Xz2WogeBnhCSQfUek.jpg   
133    4.014286  /4UMn5KmXopgpkC862Ms3t5j1sXl.jpg   
358    4.000000  /fimFFNm5LTSTaYcQEbjrFXT4rgo.jpg   
770    3.996310  /gwyJPIhCK4Xz2WogeBnhCSQfUek.jpg   
557    3.964646  /zT5UKFhq9nk97VGlBbAxlAK8UFN.jpg   
152    3.937500   /ghd5zOQnDaDW1mxO7R5fXXpZMu.jpg   
197    3.934783                              None   
492    3.900000  /8kCSlYRv9FzIsslxqoeEWfwP9JL.jpg   

                               image2 runtime  release_date original_language  
548  /36PCFtfaB4SNHVaeGZbMiMkchaT.jpg     90           1985                en  
659  /wfnMt6LGqYHcNyOfsuusw5lX3bL.jpg    153           2009                en  
606  /6VcVl48kNKvdXOZfJPdarlUGOsk.jpg    142           1969                en  
771  /49L0c5utmm0A4BkeDafM9BOhyBT.jpg    123           2000                en  
133  /1BeKi6IytbfQlQocJSvP2k6sO1a.jpg    130           2001                en  
358  /a7ptPlBghCbXkLaeSro6bNXFfIe.jpg    118           2010                en  
770  /49L0c5utmm0A4BkeDafM9BOhyBT.jpg    110           1996                en  
557  /cnKAGbX1rDkAquF2V1wVkptHDJO.jpg    151           2006                en  
152  /d8duYyyC9J5T825Hg7grmaabfxQ.jpg    136           1999                en  
197                              None     88           1943 

In [244]:
top_adventure.isnull().sum()
top_adventure.dropna(inplace=True)
top_adventure.isnull().sum()

id                   0
original_title       0
genres               0
overview             0
cast                 0
crew                 0
avg_rating           0
image1               0
image2               0
runtime              0
release_date         0
original_language    0
dtype: int64

In [245]:
import pickle
pickle.dump(top_adventure,open('top_adv.pkl','wb'))

# top international movies

In [246]:
top_foreign = movie_list[movie_list['original_language']!='en'].sort_values('avg_rating',ascending=False).head(20)
top_foreign

id                  original_title     genres  \
68     1555             L'Auberge espagnole      Drama   
3    103751     Täällä Pohjantähden alla II        NaN   
442   44674                    Après la vie      Crime   
207    2330                            Taxi     Action   
31     1284  Torrente 2: Misión en Marbella     Comedy   
758   93164             তিতাস একটি নদীর নাম     Family   
372   34326                          仁義なき戦い     Action   
347   31374                    続宮本武蔵　一乗寺の決闘      Drama   
11      110          Trois couleurs : Rouge      Drama   
607     670                            올드보이      Drama   
80   159447            С новым годом, мамы!     Comedy   
348   31408         Hababam Sınıfı Uyanıyor        NaN   
265   26497               Urmel aus dem Eis  Animation   
561    5817                      Du levande      Drama   
639   71995                 Risate di gioia     Comedy   
560   58061                    Zanna Bianca    Western   
539   56372                     Бой с Тенью     Action   
344   31297                 Shinobi no Mono     Action   
459     457                           Sissi     Comedy   
27     1269                Neues vom Wixxer     Comedy   

                                              overview  \
68   A strait-laced French student moves into an ap...   
3    Akseli Koskela returns from prison and tries t...   
442  The final installment in director Lucas Belvau...   
207  In Marseilles (France), skilled pizza delivery...   
31   For this second film in the cult comedy series...   
758  An in-depth look at the lives and struggles of...   
372  In the teeming black markets of postwar Japan,...   
347  After years on the road establishing his reput...   
11   Red This is the third film from the trilogy by...   
607  With no clue how he came to be imprisoned, dru...   
80   A continuation of a famous Russian comedy "Mam...   
348                                 No overview found.   
265  Impy's Island (German: Urmel aus dem Eis) is a...   
561  In the Swedish city of Lethe, people from diff...   
639  On New Year's Eve, an insecure, struggling act...   
560  The wolf dog, White Fang, aids a reporter, a f...   
539  Артем Колчин был одним из многих, но он хотел ...   
344  [Period covered: 1575-1681] Fist film in the S...   
459  The young Bavarian princess Elisabeth, who all...   
27   Scotland Yard investigators have to prevent th...   

                                                  cast                  crew  \
68   [Romain Duris, Audrey Tautou, Kelly Reilly, Cé...       Cédric Klapisch   
3    [Ilkka Koivula, Vera Kiiskinen, Risto Tuorila,...        Timo Koivusalo   
442  [Dominique Blanc, Gilbert Melki, Ornella Muti,...         Lucas Belvaux   
207  [Samy Naceri, Frédéric Diefenthal, Marion Coti...          Gérard Pirès   
31   [Santiago Segura, Gabino Diego, Tony Leblanc, ...       Santiago Segura   
758  [Fakrul Hasan Bairagi , Narain Chakraborty , K...         Ritwik Ghatak   
372  [Bunta Sugawara, Hiroki Matsukata, Kunie Tanak...        Kinji Fukasaku   
347  [Toshirō Mifune, Koji Tsuruta, Kaoru Yachigusa...       Hiroshi Inagaki   
11   [Irène Jacob, Jean-Louis Trintignant, Jean-Pie...  Krzysztof Kieślowski   
607  [Choi Min-sik, Yoo Ji-tae, Kang Hye-jung, Kim ...        Park Chan-wook   
80   [Alain Delon, Elizaveta Boyarskaya, Agniya Dit...      Sarik Andreasyan   
348  [Münir Özkul, Adile Naşit, Kemal Sunal, Halit ...         Ertem Eğilmez   
265  [Wigald Boning, Anke Engelke, Florian Halm, Ch...          Holger Tappe   
561  [Håkan Angser, Eric Bäckman, Patrik Anders Edg...         Roy Andersson   
639  [Anna Magnani, Totò, Ben Gazzara, Fred Clark, ...       Mario Monicelli   
560  [Franco Nero, Daniel Martín, John Steiner, Vir...           Lucio Fulci   
539  [Andrei Panin, Elena Panova, Denis Nikiforov, ...       Aleksei Sidorov   
344  [Raizô Ichikawa, Yûnosuke Itô, Shiho Fujimura,...       Satsuo Yamamoto   
459  [Romy Schneider, Karlheinz Böhm, Magda Schneid

In [247]:
top_foreign.isnull().sum()
top_foreign.dropna(inplace=True)
top_foreign.isnull().sum()

id                   0
original_title       0
genres               0
overview             0
cast                 0
crew                 0
avg_rating           0
image1               0
image2               0
runtime              0
release_date         0
original_language    0
dtype: int64

In [248]:
import pickle
pickle.dump(top_foreign,open('top_intl.pkl','wb'))

# top 90's movies

In [249]:
top_ninetys = movie_list[movie_list['release_date'].between(1989,2000)].sort_values('avg_rating',ascending=False).head(20)
top_ninetys

id                              original_title     genres  \
684   8224                                         8MM   Thriller   
674   8068                                   Desperado   Thriller   
176   2019                                 Hard Target     Action   
503    527                          Once Were Warriors      Drama   
701   8491                         Weekend at Bernie's     Comedy   
207   2330                                        Taxi     Action   
768    951                            Kindergarten Cop     Comedy   
324    306                       Beverly Hills Cop III     Action   
29    1281                                        Bean     Comedy   
771    955                      Mission: Impossible II  Adventure   
396  38499                The Howling: New Moon Rising     Horror   
757    928                   Gremlins 2: The New Batch     Comedy   
11     110                      Trois couleurs : Rouge      Drama   
359  32834                  MVP: Most Valuable Primate     Action   
705    854                                    The Mask    Romance   
305  27847  Universal Soldier III: Unfinished Business     Action   
233  25751    Children of the Corn V: Fields of Terror     Horror   
234  25752    Children of the Corn 666: Isaac's Return     Horror   
670   8011                Highlander III: The Sorcerer     Action   
571    602                            Independence Day     Action   

                                              overview  \
684  A small, seemingly innocuous plastic reel of f...   
674  A gunslinger is embroiled in a war with a loca...   
176  When a woman's father goes missing, she enlist...   
503  A drama about a Maori family lving in Auckland...   
701  Two friends are invited for a weekend to a lux...   
207  In Marseilles (France), skilled pizza delivery...   
768  Hard-edged cop John Kimble gets more than he b...   
324  Back in sunny southern California and on the t...   
29   Bean works as a caretaker at Britain's formida...   
771  With computer genius Luther Stickell at his si...   
396  Gary Brandner's horror novels come to life aga...   
757  Young sweethearts Billy and Kate move to the B...   
11   Red This is the third film from the trilogy by...   
359  Jack is a three-year-old chimpanzee who has be...   
705  When timid bank clerk Stanley Ipkiss discovers...   
305  Luc Devereaux and Veronica Roberts continue th...   
233  Six college students take a wrong turn and fin...   
234  A girl called Hannah goes back to her hometown...   
670  Starts off in the 15th century, with Connor Mc...   
571  On July 2, a giant alien mothership enters orb...   

                                                  cast                  crew  \
684  [Nicolas Cage, Joaquin Phoenix, James Gandolfi...       Joel Schumacher   
674  [Antonio Banderas, Salma Hayek, Joaquim de Alm...      Robert Rodriguez   
176  [Jean-Claude Van Damme, Arnold Vosloo, Lance H...              John Woo   
503  [Rena Owen, Temuera Morrison, Mamaengaroa Kerr...          Lee Tamahori   
701  [Andrew McCarthy, Jonathan Silverman, Catherin...          Ted Kotcheff   
207  [Samy Naceri, Frédéric Diefenthal, Marion Coti...          Gérard Pirès   
768  [Arnold Schwarzenegger, Penelope Ann Miller, P...          Ivan Reitman   
324  [Eddie Murphy, Judge Reinhold, Héctor Elizondo...           John Landis   
29   [Rowan Atkinson, Peter MacNicol, John Mills, P...             Mel Smith   
771  [Tom Cruise, Dougray Scott, Thandie Newton, Vi...              John Woo   
396  [John Ramsden, Ernest Kester, Clive Turner, Ja...          Clive Turner   
757  [Zach Galligan, Phoebe Cates, John Glover, Rob...             Joe Dante   
11   [Irène Jacob, Jean-Louis Trintignant, Jean-Pie...  Krzysztof Kieślowski   
359  [Bernie, Louie, Mac, Jamie Renée Smith, Alexa ...          Robert Vince   
705  [Jim Carrey, Cameron Diaz, Nancy Fish, Tim Bag...         Chuck Russell   
305  [Matt Battaglia, Chandra West, Roger Periard, ...     

In [250]:
top_ninetys.isnull().sum()
top_ninetys.dropna(inplace=True)
top_ninetys.isnull().sum()

id                   0
original_title       0
genres               0
overview             0
cast                 0
crew                 0
avg_rating           0
image1               0
image2               0
runtime              0
release_date         0
original_language    0
dtype: int64

In [251]:
import pickle
pickle.dump(top_ninetys,open('top_ninetys.pkl','wb'))

In [252]:
movie_list.shape

(767, 12)

In [253]:
import pickle
pickle.dump(movie_list,open('movie_list.pkl','wb'))

In [254]:
movie_list.shape

(767, 12)

In [255]:
movie_csv = pd.DataFrame(movie_list)

In [256]:
movie_list.head(1)

id      original_title  genres  \
0  100042  Dumb and Dumber To  Comedy   

                                            overview  \
0  20 years after the dimwits set out on their fi...   

                                                cast            crew  \
0  [Jim Carrey, Jeff Daniels, Rachel Melvin, Kath...  Bobby Farrelly   

   avg_rating                            image1  \
0         3.0  /fdB86tl7SKIBWZzm0CbrYjOdE1K.jpg   

                             image2 runtime  release_date original_language  
0  /saUnj5K6buVafDfLfTozDH2eqdC.jpg    110           2014                en

# COSINE SIMILARITY

In [257]:
ratings_with_id.shape

(107980, 13)

In [291]:
ratings_with_id.head(1)

userId   id  rating genres original_language  release_date  \
0       1  110     1.0  Drama                fr          1994   

           original_title runtime  \
0  Trois couleurs : Rouge     99    

                                            overview  \
0  Red This is the third film from the trilogy by...   

                                                cast                  crew  \
0  [Irène Jacob, Jean-Louis Trintignant, Jean-Pie...  Krzysztof Kieślowski   

                             image1                            image2  
0  /rVdd23QuT5rHX7lZvuAkRRUkeCZ.jpg  /AeHExfHIl70SZCea907KfEoSkfJ.jpg

users who gave minimum 30 ratings to movies......

In [258]:
x=ratings_with_id.groupby('userId').count()['rating'] > 30
litrate_user = x[x].index

out of all user's who had rated the movies...this will filtered the user's who are litrate....

In [259]:
filtered_rating = ratings_with_id[ratings_with_id['userId'].isin(litrate_user)]

movies that have minimum 30 ratings....

In [260]:
y = filtered_rating.groupby('original_title').count()['rating']>=30
famous_movies = y[y].index

out of all movies.. this will filtered the movies that have 30 or more ratings...

In [261]:
final_ratings = filtered_rating[filtered_rating['original_title'].isin(famous_movies)]

In [262]:
final_ratings.shape

(44097, 13)

In [263]:
pt = final_ratings.pivot_table(index='original_title',columns='userId',values='rating') 
pt

userId                 24     46     49     62     65     120    132    150    \
original_title                                                                  
2 Days in Paris          NaN    NaN    NaN    NaN    NaN    3.5    NaN    NaN   
2001: A Space Odyssey    2.0    NaN    NaN    3.5    NaN    2.5    NaN    3.0   
2010                     NaN    NaN    NaN    NaN    NaN    NaN    4.0    NaN   
28 Days Later            NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
28 Weeks Later           NaN    2.0    NaN    NaN    NaN    NaN    NaN    NaN   
...                      ...    ...    ...    ...    ...    ...    ...    ...   
リング                      3.0    NaN    NaN    4.0    NaN    NaN    4.0    NaN   
座頭市関所破り                  NaN    NaN    NaN    4.0    NaN    4.0    NaN    NaN   
怪獣総進撃                    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
新最佳拍檔                    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
暗戰                       NaN    NaN    NaN    NaN    3.0    NaN    NaN    NaN   

userId                 191    193    ...  10567  10568  10573  10575  10607  \
original_title                       ...                                      
2 Days in Paris          NaN    NaN  ...    NaN    NaN    NaN    5.0    3.0   
2001: A Space Odyssey    5.0    NaN  ...    NaN    3.5    NaN    3.0    NaN   
2010                     NaN    NaN  ...    NaN    NaN    NaN    NaN    NaN   
28 Days Later            NaN    3.0  ...    NaN    NaN    2.5    NaN    NaN   
28 Weeks Later           NaN    NaN  ...    2.5    NaN    1.5    NaN    1.0   
...                      ...    ...  ...    ...    ...    ...    ...    ...   
リング                      NaN    NaN  ...    2.5    4.0    4.0    NaN    NaN   
座頭市関所破り                  NaN    NaN  ...    NaN    NaN    NaN    NaN    NaN   
怪獣総進撃                    NaN    NaN  ...    NaN    4.0    NaN    NaN    2.0   
新最佳拍檔                    NaN    NaN  ...    NaN    NaN    2.5    NaN    NaN   
暗戰                       NaN    NaN  ...    NaN    NaN    NaN    NaN    4.0   

userId                 10608  10633  10637  10655  10656  
original_title                                            
2 Days in Paris          NaN    NaN    NaN    NaN    NaN  
2001: A Space Odyssey    NaN    4.0    4.0    NaN    NaN  
2010                     4.5    NaN    NaN    4.5    NaN  
28 Days Later            NaN    3.0    NaN    NaN    3.0  
28 Weeks Later           NaN    NaN    0.5    NaN    2.0  
...                      ...    ...    ...    ...    ...  
リング                      NaN    NaN    4.0    NaN    NaN  
座頭市関所破り                  NaN    NaN    3.0    NaN    NaN  
怪獣総進撃                    NaN    NaN    NaN    NaN    NaN  
新最佳拍檔                    NaN    NaN    NaN    NaN    NaN  
暗戰                       NaN    NaN    NaN    NaN    NaN  

[309 rows x 830 columns]

In [264]:
pt.fillna(0,inplace=True)
pt

userId                 24     46     49     62     65     120    132    150    \
original_title                                                                  
2 Days in Paris          0.0    0.0    0.0    0.0    0.0    3.5    0.0    0.0   
2001: A Space Odyssey    2.0    0.0    0.0    3.5    0.0    2.5    0.0    3.0   
2010                     0.0    0.0    0.0    0.0    0.0    0.0    4.0    0.0   
28 Days Later            0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
28 Weeks Later           0.0    2.0    0.0    0.0    0.0    0.0    0.0    0.0   
...                      ...    ...    ...    ...    ...    ...    ...    ...   
リング                      3.0    0.0    0.0    4.0    0.0    0.0    4.0    0.0   
座頭市関所破り                  0.0    0.0    0.0    4.0    0.0    4.0    0.0    0.0   
怪獣総進撃                    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
新最佳拍檔                    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
暗戰                       0.0    0.0    0.0    0.0    3.0    0.0    0.0    0.0   

userId                 191    193    ...  10567  10568  10573  10575  10607  \
original_title                       ...                                      
2 Days in Paris          0.0    0.0  ...    0.0    0.0    0.0    5.0    3.0   
2001: A Space Odyssey    5.0    0.0  ...    0.0    3.5    0.0    3.0    0.0   
2010                     0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0   
28 Days Later            0.0    3.0  ...    0.0    0.0    2.5    0.0    0.0   
28 Weeks Later           0.0    0.0  ...    2.5    0.0    1.5    0.0    1.0   
...                      ...    ...  ...    ...    ...    ...    ...    ...   
リング                      0.0    0.0  ...    2.5    4.0    4.0    0.0    0.0   
座頭市関所破り                  0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0   
怪獣総進撃                    0.0    0.0  ...    0.0    4.0    0.0    0.0    2.0   
新最佳拍檔                    0.0    0.0  ...    0.0    0.0    2.5    0.0    0.0   
暗戰                       0.0    0.0  ...    0.0    0.0    0.0    0.0    4.0   

userId                 10608  10633  10637  10655  10656  
original_title                                            
2 Days in Paris          0.0    0.0    0.0    0.0    0.0  
2001: A Space Odyssey    0.0    4.0    4.0    0.0    0.0  
2010                     4.5    0.0    0.0    4.5    0.0  
28 Days Later            0.0    3.0    0.0    0.0    3.0  
28 Weeks Later           0.0    0.0    0.5    0.0    2.0  
...                      ...    ...    ...    ...    ...  
リング                      0.0    0.0    4.0    0.0    0.0  
座頭市関所破り                  0.0    0.0    3.0    0.0    0.0  
怪獣総進撃                    0.0    0.0    0.0    0.0    0.0  
新最佳拍檔                    0.0    0.0    0.0    0.0    0.0  
暗戰                       0.0    0.0    0.0    0.0    0.0  

[309 rows x 830 columns]

Identifying Uclidian distance of one movies with other 663 movies.....

In [265]:
from sklearn.metrics.pairwise import cosine_similarity

In [266]:
similarity_scores = cosine_similarity(pt)

Function that identifies the 6 movies related to movie title which uclidian distance is nearer....

In [267]:
index2 = np.where(pt.index=="X-Men Origins: Wolverine")[0][0]
index2

similar_items1 = sorted(list(enumerate(similarity_scores[index2])),key=lambda x:x[1],reverse=True)[1:6]
similar_items1

[(205, 0.5086046378981148),
 (38, 0.49968421108803246),
 (210, 0.4889450455700844),
 (159, 0.4793127932840089),
 (237, 0.47629663132092015)]

In [268]:
index3 = np.where(pt.index=="X-Men Origins: Wolverine")[0][0]
index3

similar_items2 = sorted(list(enumerate(similarity_scores[index3])),key=lambda x:x[1],reverse=True)[1:16]
similar_items2

[(205, 0.5086046378981148),
 (38, 0.49968421108803246),
 (210, 0.4889450455700844),
 (159, 0.4793127932840089),
 (237, 0.47629663132092015),
 (240, 0.46773853596529125),
 (182, 0.46145243040650374),
 (62, 0.46022249208354815),
 (183, 0.4553630401879194),
 (207, 0.44112699835154867),
 (61, 0.4371955543299465),
 (261, 0.43672412266197963),
 (285, 0.4362265265473837),
 (304, 0.4312926193021354),
 (174, 0.43058999330652)]

In [269]:
pt1 = final_ratings.pivot_table(index='id',columns='userId',values='rating') 
pt1

userId  24     46     49     62     65     120    132    150    191    193    \
id                                                                             
105       NaN    1.5    NaN    NaN    NaN    NaN    NaN    NaN    NaN    2.0   
1051      NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
1091      4.0    1.5    NaN    NaN    NaN    NaN    1.0    2.0    3.0    NaN   
11        NaN    NaN    NaN    NaN    3.0    NaN    3.0    3.0    4.0    NaN   
110       5.0    NaN    4.0    NaN    5.0    3.0    3.0    4.0    5.0    5.0   
...       ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
953       NaN    NaN    5.0    3.0    5.0    NaN    3.0    NaN    3.0    NaN   
954       NaN    NaN    NaN    NaN    NaN    3.5    4.0    NaN    NaN    NaN   
955       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
97        NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
990       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   

userId  ...  10567  10568  10573  10575  10607  10608  10633  10637  10655  \
id      ...                                                                  
105     ...    NaN    NaN    3.0    NaN    NaN    NaN    NaN    NaN    NaN   
1051    ...    NaN    NaN    NaN    4.0    1.0    NaN    NaN    NaN    NaN   
1091    ...    2.5    NaN    NaN    NaN    1.0    NaN    NaN    4.0    NaN   
11      ...    NaN    4.0    NaN    4.0    2.0    NaN    NaN    4.0    NaN   
110     ...    4.0    4.0    3.5    NaN    NaN    NaN    5.0    4.5    4.0   
...     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
953     ...    5.0    NaN    3.0    NaN    NaN    NaN    NaN    2.5    3.0   
954     ...    NaN    NaN    NaN    NaN    3.0    NaN    NaN    NaN    NaN   
955     ...    NaN    NaN    NaN    NaN    5.0    NaN    NaN    NaN    NaN   
97      ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
990     ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    3.0   

userId  10656  
id             
105       NaN  
1051      NaN  
1091      NaN  
11        2.0  
110       4.0  
...       ...  
953       NaN  
954       NaN  
955       NaN  
97        NaN  
990       NaN  

[311 rows x 830 columns]

In [270]:
def recommend(movie_name):
    #index fetch
    index1 = np.where(pt.index==movie_name)[0][0]
    similar_items = sorted(list(enumerate(similarity_scores[index1])),key=lambda x:x[1],reverse=True)[1:16]
    
    data = []
    for i in similar_items:
        item = []
        temp_df = movies[movies['original_title'] == pt.index[i[0]]]
        item.extend(list(temp_df.drop_duplicates('original_title')['original_title'].values))
        item.extend(list(temp_df.drop_duplicates('original_title')['id'].values))
        item.extend(list(temp_df.drop_duplicates('original_title')['crew'].values))
        
        
        data.append(item)
        
    return data

In [271]:
pt1.fillna(0,inplace=True)
pt1

userId  24     46     49     62     65     120    132    150    191    193    \
id                                                                             
105       0.0    1.5    0.0    0.0    0.0    0.0    0.0    0.0    0.0    2.0   
1051      0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
1091      4.0    1.5    0.0    0.0    0.0    0.0    1.0    2.0    3.0    0.0   
11        0.0    0.0    0.0    0.0    3.0    0.0    3.0    3.0    4.0    0.0   
110       5.0    0.0    4.0    0.0    5.0    3.0    3.0    4.0    5.0    5.0   
...       ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
953       0.0    0.0    5.0    3.0    5.0    0.0    3.0    0.0    3.0    0.0   
954       0.0    0.0    0.0    0.0    0.0    3.5    4.0    0.0    0.0    0.0   
955       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
97        0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
990       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

userId  ...  10567  10568  10573  10575  10607  10608  10633  10637  10655  \
id      ...                                                                  
105     ...    0.0    0.0    3.0    0.0    0.0    0.0    0.0    0.0    0.0   
1051    ...    0.0    0.0    0.0    4.0    1.0    0.0    0.0    0.0    0.0   
1091    ...    2.5    0.0    0.0    0.0    1.0    0.0    0.0    4.0    0.0   
11      ...    0.0    4.0    0.0    4.0    2.0    0.0    0.0    4.0    0.0   
110     ...    4.0    4.0    3.5    0.0    0.0    0.0    5.0    4.5    4.0   
...     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
953     ...    5.0    0.0    3.0    0.0    0.0    0.0    0.0    2.5    3.0   
954     ...    0.0    0.0    0.0    0.0    3.0    0.0    0.0    0.0    0.0   
955     ...    0.0    0.0    0.0    0.0    5.0    0.0    0.0    0.0    0.0   
97      ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
990     ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    3.0   

userId  10656  
id             
105       0.0  
1051      0.0  
1091      0.0  
11        2.0  
110       4.0  
...       ...  
953       0.0  
954       0.0  
955       0.0  
97        0.0  
990       0.0  

[311 rows x 830 columns]

In [272]:
from sklearn.metrics.pairwise import cosine_similarity

In [273]:
similarity_scores1 = cosine_similarity(pt1)

In [274]:
final_movies = final_movies[['id','original_title','genres','overview','cast','crew','image1','image2','runtime','release_date','original_language']]

In [275]:
final_movielistt = pd.DataFrame(final_movies)

In [276]:
movies.isnull().sum()

belongs_to_collection       0
genres                      0
id                          0
original_language           0
release_date                0
original_title              0
runtime                     0
overview                    0
cast                        0
crew                        0
image1                    518
image2                   1216
dtype: int64

In [277]:
final_movies.shape

(4480, 11)

In [278]:
testmovie_list=movie_list[['id','avg_rating']]

In [279]:
testmovie_listright=testmovie_list.merge(final_movies,on='id',how='right')

In [280]:
testmovie_listright.shape

(4480, 12)

In [127]:
testmovie_listright.to_csv("final_ml.csv", index=False)

In [281]:
import pickle

pickle.dump(similarity_scores1,open('similarity_scores.pkl','wb'))

In [282]:
testmovie_listright[['avg_rating']] = testmovie_listright[['avg_rating']].fillna(value=1)

In [283]:
testmovie_listright.isnull().sum()

id                      0
avg_rating              0
original_title          0
genres                 52
overview                0
cast                    0
crew                   31
image1                518
image2               1216
runtime                 0
release_date            0
original_language       0
dtype: int64

In [284]:
testmovie_listright.shape

(4480, 12)

In [285]:
testmovie_listright.head(1)

id  avg_rating original_title     genres  \
0  862    3.454545      Toy Story  Animation   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   

                                                cast           crew  \
0  [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...  John Lasseter   

                             image1                            image2 runtime  \
0  /7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg  /9FBwqcd9IRruEDUrTdcaafOMKUq.jpg     81    

   release_date original_language  
0          1995                en

In [286]:
def recommend3(movie_id):
    #index fetch
    
    index1 = next((i for i, title in enumerate(pt1.index) if title == movie_id), None)
    similar_items = sorted(list(enumerate(similarity_scores1[index1])),key=lambda x:x[1],reverse=True)[1:15]
    
    data = []
    for i in similar_items:
        item = []
        temp_df = testmovie_listright[testmovie_listright['id'] == pt1.index[i[0]]]
        item.extend(list(temp_df.drop_duplicates('id')['id'].values))
        item.extend(list(temp_df.drop_duplicates('id')['original_title'].values))
        item.extend(list(temp_df.drop_duplicates('id')['image1'].values))
        data.append(item)
    print(index1)    
    return data

In [287]:
recommend3('608')

251


[['296',
  'Terminator 3: Rise of the Machines',
  '/vxiKtcxAJxHhlg2H1X8y7zcM3k6.jpg'],
 ['527', 'Once Were Warriors', None],
 ['457', 'Sissi', '/4tVV2YsrhK2hi43LlqNBRBvmsIQ.jpg'],
 ['377', 'A Nightmare on Elm Street', '/BcgmrOz69AUBRukw5LLGeYsOgo.jpg'],
 ['150', '48 Hrs.', '/8Dlcb8aDch6sUcWBc8zSSfNbPgG.jpg'],
 ['110', 'Trois couleurs : Rouge', '/rVdd23QuT5rHX7lZvuAkRRUkeCZ.jpg'],
 ['1682', 'Mosura tai Gojira', '/scvwS6k8gIW8w24UcmePQqVL10l.jpg'],
 ['923', 'Dawn of the Dead', '/mSJ20RTWWgtt3aTyn1iwNZ9fbYl.jpg'],
 ['21', 'The Endless Summer', '/zhG3P4ALlXxoFi0CGfcVsR3qmDL.jpg'],
 ['1246', 'Rocky Balboa', '/mCY5dMkSSFQufGCViI6jNUU6pXq.jpg'],
 ['25', 'Jarhead', '/aRFpBjCseFD6UmahAuLdqS7Or5q.jpg'],
 ['2791',
  'The Chronicles of Riddick: Dark Fury',
  '/4R2zrqk9KzgxU9UDH3yUUsENXR.jpg'],
 ['16', 'Dancer in the Dark', '/cwrmUIVzOuR9FdqKBfERkMHdde0.jpg'],
 ['1266', 'Street Kings', '/r0c2hR2KQsZO3vdhFQdUsnSuUef.jpg']]

In [288]:
pickle.dump(testmovie_listright,open('movies.pkl','wb'))

In [289]:
pickle.dump(pt1,open('pt1.pkl','wb'))